datafusion/src/physical_plan/expressions/approx_distinct.rs

```diff
@@ -0,0 +1,145 @@
+
+use super::format_state_name;
+use crate::error::Result;
+use crate::physical_plan::{Accumulator, AggregateExpr, PhysicalExpr};
+use crate::scalar::ScalarValue;
+use arrow::compute;
+use arrow::datatypes::DataType;
+use arrow::{
+    array::{ArrayRef, UInt64Array},
+    datatypes::Field,
+};
+use pdatastructs::hyperloglog::HyperLogLog;
+use std::any::Any;
+use std::hash::Hash;
+use std::sync::Arc;
+
+/// APPROX_DISTINCT aggregate expression
+#[derive(Debug)]
+pub struct ApproxDistinct {
+    name: String,
+    data_type: DataType,
+    nullable: bool,
+    expr: Arc<dyn PhysicalExpr>,
+}
+
+impl ApproxDistinct {
+    /// Create a new ApproxDistinct aggregate function.
+    pub fn new(
+        expr: Arc<dyn PhysicalExpr>,
+        name: impl Into<String>,
+        data_type: DataType,
+    ) -> Self {
+        Self {
+            name: name.into(),
+            expr,
+            data_type,
+            nullable: true,
+        }
+    }
+}
+
+impl AggregateExpr for ApproxDistinct {
+    /// Return a reference to Any that can be used for downcasting
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn field(&self) -> Result<Field> {
+        Ok(Field::new(
+            &self.name,
+            self.data_type.clone(),
+            self.nullable,
+        ))
+    }
+
+    fn state_fields(&self) -> Result<Vec<Field>> {
+        Ok(vec![Field::new(
+            &format_state_name(&self.name, "approx_distinct"),
+            self.data_type.clone(),
+            true,
+        )])
+    }
+
+    fn expressions(&self) -> Vec<Arc<dyn PhysicalExpr>> {
+        vec![self.expr.clone()]
+    }
+
+    fn create_accumulator(&self) -> Result<Box<dyn Accumulator>> {
+        // FIXME use switch
+        Ok(Box::new(ApproxDistinctAccumulator::<u32>::new()))
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+}
+
+#[derive(Debug)]
+struct ApproxDistinctAccumulator<T: Hash> {
+    hll: HyperLogLog<T>,
+}
+
+const REGISTERS: usize = 16_usize;
+
+impl<T> ApproxDistinctAccumulator<T>
+where
+    T: Hash,
+{
+    /// new approx_distinct accumulator
+    pub fn new() -> Self {
+        // TODO use xx_hash
+        Self {
+            hll: HyperLogLog::new(REGISTERS),
+        }
+    }
+}
+
+impl<T> Accumulator for ApproxDistinctAccumulator<T>
+where
+    T: Hash + Send + Sync + std::fmt::Debug,
+{
+    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
+        unimplemented!("not implemented")
+    }
+
+    fn update(&mut self, values: &[ScalarValue]) -> Result<()> {
+        unimplemented!("not implemented")
+    }
+
+    fn merge(&mut self, states: &[ScalarValue]) -> Result<()> {
+        unimplemented!("not implemented")
+    }
+
+    fn merge_batch(&mut self, states: &[ArrayRef]) -> Result<()> {
+        unimplemented!("not implemented")
+    }
+
+    fn state(&self) -> Result<Vec<ScalarValue>> {
+        // TODO: maybe use binary type so that merge can work?
```

Comment:
I think a binary type as the intermediate so calling `[hll::merge()`](https://docs.rs/pdatastructs/0.6.0/x86_64-pc-windows-msvc/pdatastructs/hyperloglog/struct.HyperLogLog.html#method.merge) would be a good idea